ddl/cluster.go

```diff
@@ -491,15 +494,78 @@ func finishFlashbackCluster(w *worker, job *model.Job) error {
 					return err
 				}
 			}
+			if job.IsDone() || job.IsSynced() {
+				gcSafePoint, err := gcutil.GetGCSafePoint(sess)
+				if err != nil {
+					return err
+				}
+				if err = UpdateFlashbackHistoryTSRanges(t, flashbackTS, t.StartTS, gcSafePoint); err != nil {
+					return err
+				}
+			}
 			if err = t.SetFlashbackClusterJobID(0); err != nil {
 				return err
 			}
 		}
 		return nil
 	})
 	if err != nil {
+		logutil.BgLogger().Warn("Finish flashback cluster meets error", zap.Error(err))
 		return err
 	}
 
 	return nil
 }
+
+// CheckFlashbackHistoryTSRange checks flashbackTS overlapped with history time ranges or not.
+func CheckFlashbackHistoryTSRange(m *meta.Meta, flashbackTS uint64) error {
+	tsRanges, err := m.GetFlashbackHistoryTSRange()
+	if err != nil {
+		return err
+	}
+	for _, tsRange := range tsRanges {
+		if tsRange.StartTS <= flashbackTS && flashbackTS <= tsRange.EndTS {
+			return errors.Errorf("FlashbackTs overlapped, old range: [%s, %s], flashbackTS: %s",
+				oracle.GetTimeFromTS(tsRange.StartTS), oracle.GetTimeFromTS(tsRange.EndTS), oracle.GetTimeFromTS(flashbackTS))
+		}
+	}
+	return nil
+}
+
+// UpdateFlashbackHistoryTSRanges insert [startTS, endTS] into FlashbackHistoryTSRange.
+func UpdateFlashbackHistoryTSRanges(m *meta.Meta, startTS uint64, endTS uint64, gcSafePoint uint64) error {
+	tsRanges, err := m.GetFlashbackHistoryTSRange()
+	if err != nil {
+		return err
+	}
+	if len(tsRanges) != 0 && tsRanges[len(tsRanges)-1].EndTS >= endTS {
+		// It's impossible, endTS should always greater than all TS in history TS ranges.
+		return errors.Errorf("Invalid flashback ts range, last flashback end time: %s, now: %s",
+			oracle.GetTimeFromTS(tsRanges[len(tsRanges)-1].EndTS), oracle.GetTimeFromTS(endTS))
+	}
+
+	newTsRange := make([]meta.TSRange, 0, len(tsRanges))
+
+	for _, tsRange := range tsRanges {
+		if tsRange.EndTS < gcSafePoint {
+			continue
+		}
+		if startTS > tsRange.EndTS {
+			// tsRange.StartTS < tsRange.EndTS < startTS.
+			// We should keep tsRange in slices.
+			newTsRange = append(newTsRange, tsRange)
+		} else if startTS < tsRange.StartTS {
+			// startTS < tsRange.StartTS < tsRange.EndTS.
```

Comment:
```
// startTS < tsRange.StartTS < tsRange.EndTS < endTS(= new startTS, thus the max)
```

Want something for `endTS`, too.

Reply:
checked in https://github.com/pingcap/tidb/pull/37737/files#diff-ce4a909a18db5fdccf707860be14b509161216c5041aa76fff6110dfefe9bf86R542